python/lsst/ts/ess_sensors/rpi_serial_hat.py

```diff
@@ -0,0 +1,383 @@
+        """
+        with self._lock:
+            try:
+                self._ser
+            except NameError:
+                self._message("Cannot open. Serial port does not exist.")
+                raise RuntimeWarning(
+                    f"{self.name}: Could not open. Serial port does not exist."
+                )
+            else:
+                if not self._ser.is_open:
+                    try:
+                        await self._ser.open()
+                        self._message("Serial port opened.")
+                    except serial.SerialException as e:
+                        self._message("Serial port open failed.")
+                        raise e
+                else:
+                    self._message("Port already open!")
+
+    async def close(self) -> None:
+        """Close serial communications port.
+
+        Raises
+        ------
+        IOError if serial communications port fails to close.
+        """
+        with self._lock:
+            try:
+                self._ser
+            except NameError:
+                self._message("Cannot close. Serial port does not exist.")
+                raise RuntimeWarning(f"{self.name}: Could not close the serial port!")
+            else:
+                if self._ser.is_open:
+                    await self._ser.close()
+                    self._message("Serial port closed.")
+                else:
+                    self._message("Serial port already closed.")
+
+    def readline(self) -> str:
+        r"""Read a line of ASCII string data from the serial port.
+
+        Returns
+        -------
+        name, error, resp : `tuple`
+        name : `str`
+            The name of the device.
+        error : `str`
+            Error string.
+            'OK' = No error
+            'Non-ASCII data in response.'
+            'Timed out with incomplete response.'
+        resp : `str`
+            Response read from the VCP.
+            Includes terminator string if there is one.
+            May be returned empty if nothing was received, partial if the
+            readline was started during device reception or partial if
+            the read was terminated during reception of data due to timeout.
+
+        Notes
+        -----
+        Reads device incoming buffer one character at a time until any of
+        read_timeout, terminator or line_size constraints are met.
+        Implements read timeout, returning if read_timeout is exceeded.
+        Tests for decode error, enforcing ASCII data only.
+        Optionally implements ASCII line terminator string (Commonly '\r\n'),
+        returning when terminator string is read.
+        Optionally implements line size, returning when line_size is met if a
+        terminator string is empty.
+        """
+
+        resp: str = ""
+        err: str = "OK"
+        await self._rpi_pin_state(self._pin_dirn, RpiSerialHat.STATE_DIRN_RX)
+        with self._lock:
+            while not resp.endswith("\r\n"):
```

Comment:
You have hard-coded the terminator here, despite making it a property. Please look for other such instances and add a unit test for nonstandard terminators.

Reply:
I agree. This will be done in DM-33072.

Reply:
Done